New: Disallow inline rule configuration option

tofumatt opened the pull request:
Closes #3901.

This adds a `disallowInlineConfig` option, off by default, that disables
inline configuration in comments of a particular file.

This disable:

* `/*eslint-disable*/`
* `/*eslint-enable*/`
* `/*global*/`
* `/*eslint*/`
* `/*eslint-env*/`

tofumatt commented:
Hi there! I finally got around to writing this patch; hope I got things right. Please let me know if there's anything I should do to fix up the code to get the PR accepted. Cheers! :smile:

lo1tuma commented on docs/user-guide/command-line-interface.md:
```diff
@@ -315,6 +317,22 @@ Example:
+* `/*eslint-enable*/`
+* `/*global*/`
+* `/*eslint*/`
+* `/*eslint-env*/`
```
What about `// eslint-disable-line` comments?

tofumatt replied:
It should ignore those too (it prevents `modifyConfigsFromComments` from running at all), but I didn't document that or add tests for them. I'll do that and update the PR.

tofumatt replied:
Added the fix (it's in a separate commit but I can squash before merge, figured I'd wait until the end of review beforehand :smile:).

Thanks for catching that. :)

gyandeeps commented:
I see that the issue has not been marked accepted. Are we still working on the what/how for the functionality?

* Need to add unit test for options file.

tofumatt commented:
Either way I needed the functionality for myself, so I submitted the PR as part of the discussion.

From what I read though there was functionality discussed in my issue that was separate from what I wanted and it was said to be out-of-scope for my issue. After that discussion died down, and I figured code was a better ping than just a comment. 😀

I'll make a note to add a unit test for the options file, thank you.

tofumatt commented:
Added the unit tests for `--disableInlineConfig` option.

Happy to rename that of course, but it's what made sense to me.

nzakas commented on docs/user-guide/command-line-interface.md:
````diff
@@ -67,7 +67,9 @@ Miscellaneous:
   -h, --help                 Show help
   -v, --version              Outputs the version number
-  ```
+  --disableInlineConfig      Prevent comments from changing eslint rules -
````
Command line options need to be dash-cased. Maybe `no-inline-config`?

tofumatt replied:
Sure thing; I’ll patch that up in a PR later tonight.

nzakas commented on lib/eslint.js:
```diff
@@ -413,6 +413,7 @@ function prepareConfig(config) {
         globals: util.mergeConfigs({}, config.globals),
         env: util.mergeConfigs({}, config.env || {}),
         settings: util.mergeConfigs({}, config.settings || {}),
+        disableInlineConfig: config.disableInlineConfig || false,
```
I'm not sure adding `disableInlineConfig` to `config` is a good idea. That means people can do that inside of their `.eslintrc` files, and we need to think a lot harder about exposing that.

tofumatt replied:
Ah, that makes sense actually. As long as I can enable it via the CLI and using CLIEngine I’m happy; honestly I was a bit lost on where to add things and may have went a bit trigger-happy.

tofumatt replied:
Actually, now I'm a bit confused. Sorry but I'm finding it a bit hard to figure out how to add an option for `CLIEngine` but not as a config option.

Where should I add this option without allowing `.eslintrc` files from adding it?

nzakas replied:
It would probably need to be a new argument to `verify`

tofumatt commented:
Okay, this is ready again for review, I believe it's without the option (so it can't be enabled in a config file) but it's activated either via `--disable-inline-config=true` or `new ESLint.CLIEngine({disableInlineConfig: true});`

My commit message fits the check and my tests pass; not sure why Travis is complaining. Maybe re-run the test?

ilyavolodin commented:
Travis is complaining because commit messages should contain `(fixes #)` on the first line

tofumatt commented:
Oh. The error on travis complains about the LAST line.

I'll fix that up in a bit.

platinumazure commented:
@tofumatt Actually, it just says the "commit message must end" with one of those. But that is confusing.

Sometimes people differentiate between a "commit summary" and a "commit body", which refer to the first line of a commit message and everything else, respectively. If you want to change the message to use either that language, or just say "First line of commit message", feel free to open an issue and/or PR.

tofumatt commented:
Gotcha, good to know. 

Yeah, I'll see about a separate pr for that, thanks!

- Matthew Riley MacPherson (Sent from mobile)

> On 25 Oct 2015, at 17:24, Kevin Partington <notifications@github.com> wrote:
> 
> @tofumatt Actually, it just says the "commit message must end" with one of those. But that is confusing.
> 
> Sometimes people differentiate between a "commit summary" and a "commit body", which refer to the first line of a commit message and everything else, respectively. If you want to change the message to use either that language, or just say "First line of commit message", feel free to open an issue and/or PR.
> 
> —
> Reply to this email directly or view it on GitHub.
>

tofumatt commented:
Cool, fixed the commit message and it's as green as grass now!

nzakas commented on lib/options.js:
```diff
@@ -193,6 +193,12 @@ module.exports = optionator({
             description: "Outputs the version number"
+        },
+        {
+            option: "disable-inline-config",
```
We should do this in the same manner as the `eslintrc` option. Name the option "inline-config", set default to `true`, then optionator will make that `no-inline-config`.

nzakas commented on lib/eslint.js:
```diff
@@ -620,9 +620,11 @@ module.exports = (function() {
      * @returns {Object[]} The results as an array of messages or null if no messages.
      */
-    api.verify = function(textOrSourceCode, config, filename, saveState) {
+    api.verify = function(textOrSourceCode, config, filename, saveState, disableInlineConfig) {
```
I think we've hit peak number of arguments here. Since this is exposed publicly, I'd prefer not to add a new argument (we're already pushing it, and `saveState` is dangerous to trigger outside of a test.

@eslint/eslint-team other thoughts on how this could work?

gyandeeps replied:
I think we can accept a `options` object here which contains all the current attributes plus the new attribute.
T make it backward compatible, we can check if the first argument is a `string` or `instance of SourceCode` then consider it a old style otherwise new style(which is a object).

tofumatt replied:
That makes sense to me, but that is a different issue/PR from this.

Seeing as my patch works, could we look at merging it then filing that as a separate issue? :smile:

nzakas replied:
Yeah, let's do that separately. What I'd say is left make the third argument (currently `filename`) into the new options object and let it have `filename`, `saveState`, and whatever else.

Then, @tofumatt, we'll ask you to rebase on top of that change for this PR.

tofumatt replied:
That sounds good to me! Thanks :-)

caridy commented:
@tofumatt do you need any help to get this ready?

tofumatt commented:
@caridy To my mind my patch is ready, but there was talk about waiting for another refactor for me to rebase onto.

I was going to ask though, actually: can we get this change merged now and do the refactor later? My patch is ready now and it seems to make more sense to refactor afterward. Plus Mozilla would love to have this feature in for our new add-on validator :smile: 

Thoughts about merging this in?

nzakas commented:
No, we can't merge it until the other changes are done. I don't want it going out in this form, and a refactor would be a breaking change because `verify()` is part of the public API. 

I just filed an issue for the refactor: https://github.com/eslint/eslint/issues/4395. Do you want to take a stab at that?

tofumatt commented:
Gotcha, fair enough re: changing the public API.

I can get back to you about taking that issue tomorrow or Friday at the latest. I’d like to as I’m interested in seeing this get merged, but don’t want to commit to it just yet while I sort out some other work I need to get finished first.

I’ll set a reminder to come back to it Friday at the latest, cheers :-)

nzakas commented:
I'll see if I can do it quickly...

nzakas commented:
https://github.com/eslint/eslint/pull/4398

tofumatt commented:
Howdy,

I've rebased on top of the latest master, including the fix in #4398. I switched to using `inline-config` defaulting to `true` as requested.

I think this covers everything, but a fair amount feels like it changed so please let me know if I missed anything.

Otherwise should be all good to go! :smile:

gyandeeps commented on docs/user-guide/command-line-interface.md:
````diff
@@ -69,7 +69,9 @@ Miscellaneous:
   -h, --help                 Show help
   -v, --version              Outputs the version number
-  ```
+  --no-inline-config         Prevent comments from changing eslint rules -
````
I think this should be `--inline-config` since default is `false`.

tofumatt replied:
You are correct, but the default is true. I guess I figure people will want to use `—no-inline-config` on the command line as it is prettier, but I’ll fix the docs.

tofumatt replied:
Actually, `inlineConfig` is `true` by default, but I think specifying it as `--no-inline-config` is the nicer API (like `--no-color`, above).

gyandeeps replied:
`--no-inline-config` is correct if by default its `true` but your option description says `false`. Thats why I am trying to make sure what it is?

tofumatt replied:
`inlineConfig` is `true` by default: https://github.com/eslint/eslint/pull/4186/files#diff-763aa8deb604d725a74f8304903c1422R205

nzakas replied:
This is correct.

gyandeeps commented on lib/cli-engine.js:
```diff
@@ -95,7 +95,8 @@ var defaultOptions = {
         cacheFile: ".eslintcache",
-        fix: false
+        fix: false,
+        inlineConfig: true
```
Above I saw the default was `false`, Am I missing something?

tofumatt replied:
`--no-inline-config` is `false`, meaning `inlineConfig` is `true`. I used `--no-inline-config` in the command line docs as it's a nicer, more straightforward API. But internally the behaviour is `inlineConfig` is always true/on unless explicitly marked as `false` by either the CLI or a `CLIEngine` config.

nzakas replied:
I think this should be `allowInlineConfig` as well

nzakas commented on docs/developer-guide/nodejs-api.md:
```diff
@@ -31,6 +31,7 @@ The most important method on `linter` is `verify()`, which initiates linting of
 * `options` - (optional) Additional options for this run.
     * `filename` - (optional) the filename to associate with the source code.
     * `saveState` - (optional) set to true to maintain the internal state of `linter` after linting (mostly used for testing purposes).
+    * `inlineConfig` - (optional) set to `false` to disable inline comments from changing eslint rules.
```
Probably `allowInlineConfig`, to indicate is a Boolean rather than an object.

tofumatt replied:
Oh, that’s a good point. I’ll fix and push in a few minutes.

tofumatt commented:
Pushed a version that uses `allowInlineConfig` internally but still uses `--no-inline-config` at the command line.

nzakas commented on lib/eslint.js:
```diff
@@ -608,9 +608,11 @@ module.exports = (function() {
+     *      an object, then it has "filename", "saveState", and "allowInlineConfig" properties.
      * @param {boolean} [saveState] Indicates if the state from the last run should be saved.
      *      Mostly useful for testing purposes.
+     * @param {boolean} [allowInlineConfig] Allow/disallow inline comments' ability to change config once it is set. Defaults to true if not supplied.
```
`[filenameOrOptions.allowInlineConfig]` (since it doesn't exist as a named argument)

nzakas commented on tests/lib/cli.js:
```diff
@@ -633,6 +633,30 @@ describe("cli", function() {
 
+    describe("when passed --no-inline-config", function() {
+
+        it("should pass allowInlineConfig:false to CLIEngine and ignore inline config when executing on text", function() {
```
This isn't actually what the test is doing. Looks like previous changes were reverted?

tofumatt replied:
Oops. I think I misunderstand the test then.

It's passing `--no-inline-config` which sets `allowInlineConfig` to `false`, and then the linting exists with a non-zero because it finds a rule violation that would normally be ignored as it's on a `// eslint-disable-line` line, but the `allowInlineConfig = false` means that comment config is ignored.

nzakas replied:
The description is what you should be doing - making sure the correct options are passed to CLIEngine, but you're testing that via side effects instead of direct inspection, which is fragile. If you look towards the end of this file, you'll see an example of how to test this better.

tofumatt commented:
Okay, I think I fixed the test, let me know if I did things right. I'll admit these new ones made less sense to me, so I hope I did things alright.

nzakas commented on tests/lib/cli.js:
```diff
@@ -633,6 +633,75 @@ describe("cli", function() {
+                "./logging": log
+            });
+
+            var exitCode = localCLI.execute("--no-inline-config .");
```
You don't have to check exit code here, it's inconsequential.

nzakas commented on tests/lib/cli-engine.js:
```diff
@@ -1854,4 +1854,35 @@ describe("CLIEngine", function() {
+
+    describe("when evaluating code with comments to change config when allowInlineConfig is disabled", function() {
+
+        it("should report a violation for disabling rules", function() {
```
Can you also add a test for the opposite? Make sure without the option that inline comments still work

tofumatt replied:
Sure thing; extra test (and removing the superfluous exit code check) arriving in about five minutes! :smile:

tofumatt commented:
Fixes pushed.

nzakas commented:
Looks good! Thanks for your patience.

caridy commented:
awesome, thanks for pushing this forward @tofumatt.

tofumatt commented:
Yay! Likewise: thanks for the help with the PR :smile:
